R/joins.R

```diff
@@ -55,25 +46,36 @@ join_dt <- function(op) {
 }
 
 #' @rdname join.tbl_dt
-inner_join.data.table <- join_dt({merge(x, y, by = by$x, allow.cartesian = TRUE)})
+inner_join.data.table <- join_dt({merge(x, y, by.x = by$x, by.y = by$y, all = FALSE, allow.cartesian = TRUE)})
 
 #' @rdname join.tbl_dt
-left_join.data.table  <- join_dt({merge(x, y, by = by$x, all.x = TRUE, allow.cartesian = TRUE)})
+left_join.data.table <- join_dt({merge(x, y, by.x = by$x, by.y = by$y, all.x = TRUE, allow.cartesian = TRUE)})
 
 #' @rdname join.tbl_dt
-right_join.data.table  <- join_dt(merge(x, y, by = by$x, all.y = TRUE, allow.cartesian = TRUE))
+right_join.data.table <- join_dt({merge(x, y, by.x = by$x, by.y = by$y, all.y = TRUE, allow.cartesian = TRUE)})
 
 #' @rdname join.tbl_dt
 semi_join.data.table  <- join_dt({
-  # http://stackoverflow.com/questions/18969420/perform-a-semi-join-with-data-table
-  w <- unique(x[y, which = TRUE, allow.cartesian = TRUE])
-  w <- w[!is.na(w)]
-  x[w]
+   y <- as.data.table(y)
```

Comment:
Without `copy = TRUE` we still would get:

```
> library(dplyr)
library(dplyr)
dt <- data.table::data.table(x = 1:3, y = 3:1)
df <- data.frame(x = 3:1, z = 1:3)
anti_join(dt, df)
Joining, by = "x"
Error in `[.data.frame`(y, , by_y, with = FALSE) : 
  unused argument (with = FALSE)
```

This could be avoided with `y <- as.data.table(y)`